src/HIE/Bios/Cradle.hs

```diff
@@ -124,21 +133,63 @@ noneCradle cur_dir =
     { cradleRootDir = cur_dir
     , cradleOptsProg = CradleAction
         { actionName = "default"
-        , getDependencies = return []
-        , getOptions = const $ return CradleNone
+        , runCradle = const $ return CradleNone
+        }
+    }
+
+---------------------------------------------------------------
+-- The multi cradle selects a cradle based on the filepath
+
+multiCradle :: FilePath -> [(FilePath, CradleConfig)] -> Cradle
+multiCradle cur_dir cs =
+  Cradle
+    { cradleRootDir = cur_dir
+    , cradleOptsProg = CradleAction
+        { actionName = "multi"
+        , runCradle = canonicalizePath >=> multiAction cur_dir cs
         }
     }
 
+multiAction :: FilePath
+            -> [(FilePath, CradleConfig)]
+            -> FilePath
+            -> IO (CradleLoadResult ComponentOptions)
+multiAction cur_dir cs cur_fp = selectCradle =<< canonicalizeCradles
+
+  where
+    err_msg = unlines $ ["Multi Cradle: No prefixes matched"
+                      , "pwd: " ++ cur_dir
+                      , "filepath" ++ cur_fp
+                      , "prefixes:"
+                      ] ++ [show (pf, cradleType cc) | (pf, cc) <- cs]
+
+    -- Canonicalize the relative paths present in the multi-cradle and
+    -- also order the paths by most specific first. In the cradle selection
+    -- function we want to choose the most specific cradle possible.
+    canonicalizeCradles :: IO [(FilePath, CradleConfig)]
+    canonicalizeCradles =
+      sortBy (\(p1, _) (p2, _) ->
+        if p1 `isPrefixOf` p2
+          then GT
+          else LT) <$> mapM (\(p, c) -> (,c) <$> (canonicalizePath (cur_dir </> p))) cs
```

Comment:
Implementation looks a bit complex, imo.
SortBy could be something like:
`sortOn (Down . compare . fst)` but not a blocker

Reply:
I think it's a good point. I'll change the implementation. I didn't consider that the `Ord` instance for `FilePath` would do the right thing here.